README.md
# files_external_s3

[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=owncloud_files_external_s3&metric=alert_status)](https://sonarcloud.io/dashboard?id=owncloud_files_external_s3)
[![Security Rating](https://sonarcloud.io/api/project_badges/measure?project=owncloud_files_external_s3&metric=security_rating)](https://sonarcloud.io/dashboard?id=owncloud_files_external_s3)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=owncloud_files_external_s3&metric=coverage)](https://sonarcloud.io/dashboard?id=owncloud_files_external_s3)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

## [Unreleased]

-

## [2.1.0] - 2022-12-01

### Changed

- [#1007](https://github.com/owncloud/files_external_s3/pull/1007) - Feat: performance optimizations


## [2.0.0] - 2022-07-05

### Changed

- [#39387](https://github.com/owncloud/core/issues/39387) - Update guzzle major version to 7
- This version requires ownCloud 10.11.0 or above

## 1.0.0 - 2019-09-25

- Moved from core into a separate app - [#2](https://github.com/owncloud/files_external_s3/pull/2)
- Updated AWS SDK library to 3.110.3

[Unreleased]: https://github.com/owncloud/files_external_s3/compare/v2.1.0...master
[2.1.0]: https://github.com/owncloud/files_external_s3/compare/v2.0.0...v2.1.0
[2.0.0]: https://github.com/owncloud/files_external_s3/compare/v1.0.0...v2.0.0


tests/drone/run.sh
#!/usr/bin/env bash
set -xeo pipefail

if [[ "$(pwd)" == "$(cd "$(dirname "$0")"; pwd -P)" ]]; then
  echo "Can only be executed from project root!"
  exit 1
fi

declare -x COVERAGE
[[ -z "${COVERAGE}" ]] && COVERAGE="false"

readonly BASE_DIR="$(pwd)"

wait_for_storage() {
  local storage_type="${1}"
  case "${storage_type}" in
    ceph)
        wait-for-it -t 120 ceph:80
      ;;
    scality)
        wait-for-it -t 120 scality:8000
      ;;
    minio)
        wait-for-it -t 120 minio:9000
      ;;
    *)
      echo "Unknown service \"${storage_type}\"!"
      exit 1
      ;;
  esac
}

main () {
  # wait for storage to be ready
  wait_for_storage "${STORAGE}"

  # prepare config
  cp "tests/drone/configs/config.${STORAGE}.php" "tests/unit/config.php"

  # go to server root dir
  core_path="$(dirname "$(dirname "${BASE_DIR}")")"
  cd "${core_path}"

  # enable apps
  php occ config:app:set core enable_external_storage --value=yes
  php occ app:enable files_external
  php occ app:enable files_external_s3

  # run unit tests
  if [[ "${COVERAGE}" == "true" ]]; then
    phpdbg -d memory_limit=4096M -rr ./lib/composer/bin/phpunit --configuration "${BASE_DIR}/phpunit.xml"
  else
    ./lib/composer/bin/phpunit --configuration "${BASE_DIR}/phpunit.xml"
  fi

}

main

tests/unit/bootstrap.php
<?php
require_once __DIR__ . '/../../../../tests/bootstrap.php';


tests/unit/Amazons3Test.php
<?php
/**
 * @author Joas Schilling <coding@schilljs.com>
 * @author Michael Gapczynski <GapczynskiM@gmail.com>
 * @author Morris Jobke <hey@morrisjobke.de>
 * @author Robin Appelman <icewind@owncloud.com>
 * @author Robin McCorkell <robin@mccorkell.me.uk>
 * @author Thomas Müller <thomas.mueller@tmit.eu>
 *
 * @copyright Copyright (c) 2018, ownCloud GmbH
 * @license AGPL-3.0
 *
 * This code is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License, version 3,
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 *
 */

namespace OCA\Files_External\Tests\Storage;

use OCA\FilesExternalS3\Storage\AmazonS3;

/**
 * Class Amazons3Test
 *
 * @group DB
 *
 * @package OCA\Files_External\Tests\Storage
 */
class Amazons3Test extends \Test\Files\Storage\Storage {
	private $config;

	protected function setUp(): void {
		parent::setUp();

		$this->config = include __DIR__.'/config.php';
		if (! \is_array($this->config) or ! $this->config['run']) {
			$this->markTestSkipped('AmazonS3 backend not configured');
		}
		$this->instance = new AmazonS3($this->config);
	}

	protected function tearDown(): void {
		if ($this->instance) {
			$this->instance->rmdir('');
		}

		parent::tearDown();
	}

	/**
	 * @dataProvider directoryProvider
	 */
	public function testDirectories($directory) {
		$this->markTestSkipped('S3 doesn\'t handle directories correctly - see issue 7');
	}

	public function testRecursiveRmdir() {
		$this->markTestSkipped('S3 doesn\'t handle directories correctly - see issue 7');
	}

	public function testRmdirEmptyFolder() {
		$this->markTestSkipped('S3 doesn\'t handle directories correctly - see issue 7');
	}

	public function testRecursiveUnlink() {
		$this->markTestSkipped('S3 doesn\'t handle directories correctly - see issue 7');
	}

	public function testRenameDirectory() {
		$this->markTestSkipped('S3 doesn\'t handle directories correctly - see issue 7');
	}

	public function testRenameOverWriteDirectory() {
		$this->markTestSkipped('S3 doesn\'t handle directories correctly - see issue 7');
	}

	public function testRenameOverWriteDirectoryOverFile() {
		$this->markTestSkipped('S3 doesn\'t handle directories correctly - see issue 7');
	}

	public function testCopyDirectory() {
		$this->markTestSkipped('S3 doesn\'t handle directories correctly - see issue 7');
	}

	public function testCopyOverWriteDirectory() {
		$this->markTestSkipped('S3 doesn\'t handle directories correctly - see issue 7');
	}

	public function testStat() {
		$this->markTestSkipped('S3 doesn\'t update the parents folder mtime');
	}
}


appinfo/app.php
<?php
/**
 * @author Thomas Müller <thomas.mueller@tmit.eu>
 *
 * @copyright Copyright (c) 2018, ownCloud GmbH.
 * @license GPL-2.0
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

require_once \dirname(__DIR__) . '/vendor/autoload.php';

$app = new \OCA\FilesExternalS3\AppInfo\Application();


lib/Auth/AccessKey.php
<?php
/**
 * @author Robin McCorkell <robin@mccorkell.me.uk>
 * @author Vincent Petry <pvince81@owncloud.com>
 *
 * @copyright Copyright (c) 2018, ownCloud GmbH.
 * @license GPL-2.0
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

namespace OCA\FilesExternalS3\Auth;

use OCP\Files\External\Auth\AuthMechanism;
use OCP\Files\External\DefinitionParameter;
use OCP\IL10N;

/**
 * Amazon S3 access key authentication
 */
class AccessKey extends AuthMechanism {
	public const SCHEME_AMAZONS3_ACCESSKEY = 'amazons3_accesskey';

	public function __construct(IL10N $l) {
		$this
			->setIdentifier('amazons3::accesskey')
			->setScheme(self::SCHEME_AMAZONS3_ACCESSKEY)
			->setText($l->t('Access key'))
			->addParameters([
				(new DefinitionParameter('key', $l->t('Access key'))),
				(new DefinitionParameter('secret', $l->t('Secret key')))
					->setType(DefinitionParameter::VALUE_PASSWORD),
			]);
	}
}


lib/AppInfo/Application.php
<?php
/**
 * @author Thomas Müller <thomas.mueller@tmit.eu>
 *
 * @copyright Copyright (c) 2018, ownCloud GmbH.
 * @license GPL-2.0
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

namespace OCA\FilesExternalS3\AppInfo;

use OCA\FilesExternalS3\Auth\AccessKey;
use OCA\FilesExternalS3\Backend\AmazonS3;
use OCP\AppFramework\App;
use OCP\Files\External\Config\IAuthMechanismProvider;
use OCP\Files\External\Config\IBackendProvider;

/**
 * @package OCA\FilesExternalS3\AppInfo
 */
class Application extends App implements IBackendProvider, IAuthMechanismProvider {
	public function __construct(array $urlParams = []) {
		parent::__construct('files_external_s3', $urlParams);

		$container = $this->getContainer();

		/** @var \OC\Server $server */
		$server= $container->getServer();
		/* @phan-suppress-next-line PhanUndeclaredMethod */
		$backendService = $server->getStoragesBackendService();
		$backendService->registerBackendProvider($this);
		$backendService->registerAuthMechanismProvider($this);
	}

	/**
	 * @{inheritdoc}
	 */
	public function getBackends() {
		$container = $this->getContainer();

		$backends = [
			$container->query(AmazonS3::class)
		];
		return $backends;
	}

	/**
	 * @{inheritdoc}
	 */
	public function getAuthMechanisms() {
		$container = $this->getContainer();
		return [
			$container->query(AccessKey::class)
		];
	}
}


appinfo/Migrations/Version20190905111710.php
<?php
/**
 * @author Sujith Haridasan <sharidasan@owncloud.com>
 *
 * @copyright Copyright (c) 2019, ownCloud GmbH
 * @license AGPL-3.0
 *
 * This code is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License, version 3,
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 *
 */

namespace OCA\files_external_s3\Migrations;

use OCP\IDBConnection;
use OCP\Migration\ISqlMigration;

/**
 * Replace the storage backend amazons3 to files_external_s3.
 * Reason for this is post 10.2.1 version of core the changes have been
 * moved to files_external_s3 app. So kindly install and enable the app
 * to get the data, after upgrade.
 */
class Version20190905111710 implements ISqlMigration {
	/**
	 * @param IDBConnection $connection
	 * @return array|string
	 */
	public function sql(IDBConnection $connection) {
		$qb = $connection->getQueryBuilder();
		$qb->update('external_mounts')
			->set(
				'storage_backend',
				$qb->expr()->literal('files_external_s3')
			)
			->where(
				$qb->expr()->eq('storage_backend', $qb->expr()->literal('amazons3'))
			);
		$qb->execute();

		return $qb->getSQL();
	}
}


appinfo/Migrations/Version20190911123154.php
<?php
/**
 * @author Sujith Haridasan <sharidasan@owncloud.com>
 *
 * @copyright Copyright (c) 2019, ownCloud GmbH
 * @license AGPL-3.0
 *
 * This code is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License, version 3,
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 *
 */

namespace OCA\files_external_s3\Migrations;

use OCP\Migration\ISimpleMigration;
use OCP\Migration\IOutput;

/**
 * Update amazons3 in the appconfig to files_external_s3
 */
class Version20190911123154 implements ISimpleMigration {
	/**
	 * @param IOutput $out
	 */
	public function run(IOutput $out) {
		$userMountBackends = \OC::$server->getConfig()->getAppValue('files_external', 'user_mounting_backends', '');
		if ($userMountBackends !== '') {
			$mountBackendArray = \explode(',', $userMountBackends);
			/**
			 * Search for amazons3, and replace it to files_external_s3.
			 */
			foreach ($mountBackendArray as $key => $mountBackend) {
				if ($mountBackend === 'amazons3') {
					$mountBackendArray[$key] = 'files_external_s3';
					break;
				}
			}
			$userMountBackends = \implode(',', $mountBackendArray);
			\OC::$server->getConfig()->setAppValue('files_external', 'user_mounting_backends', $userMountBackends);
		}
	}
}


lib/Storage/AmazonS3.php
<?php
/**
 * @author André Gaul <gaul@web-yard.de>
 * @author Arthur Schiwon <blizzz@arthur-schiwon.de>
 * @author Bart Visscher <bartv@thisnet.nl>
 * @author Christian Berendt <berendt@b1-systems.de>
 * @author Christopher T. Johnson <ctjctj@gmail.com>
 * @author Johan Björk <johanimon@gmail.com>
 * @author Jörn Friedrich Dreyer <jfd@butonic.de>
 * @author Martin Mattel <martin.mattel@diemattels.at>
 * @author Michael Gapczynski <GapczynskiM@gmail.com>
 * @author Morris Jobke <hey@morrisjobke.de>
 * @author Philipp Kapfer <philipp.kapfer@gmx.at>
 * @author Robin Appelman <icewind@owncloud.com>
 * @author Robin McCorkell <robin@mccorkell.me.uk>
 * @author Thomas Müller <thomas.mueller@tmit.eu>
 * @author Vincent Petry <pvince81@owncloud.com>
 *
 * @copyright Copyright (c) 2018, ownCloud GmbH.
 * @license GPL-2.0
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

namespace OCA\FilesExternalS3\Storage;

use Aws\Handler\GuzzleV6\GuzzleHandler;
use Aws\S3\Exception\S3Exception;
use Aws\S3\S3Client;
use DateTimeInterface;
use GuzzleHttp\Handler\CurlMultiHandler;
use GuzzleHttp\Middleware;
use Icewind\Streams\IteratorDirectory;
use OCP\Files\Storage\StorageAdapter;
use OCP\Files\StorageNotAvailableException;
use OCP\ILogger;
use OCP\ITempManager;
use Psr\Http\Message\RequestInterface;

class AmazonS3 extends StorageAdapter {

	/**
	 * @var \Aws\S3\S3Client
	 */
	private $connection;
	/**
	 * @var string
	 */
	private $bucket;
	/**
	 * @var array
	 */
	private static $tmpFiles = [];
	/**
	 * @var array
	 */
	private $params;
	/**
	 * @var bool
	 */
	private $test = false;
	/**
	 * @var int
	 */
	private $timeout = 15;
	/**
	 * @var int in seconds
	 */
	private $rescanDelay = 10;

	/** @var string */
	private $id;

	/** @var ILogger */
	private $logger;

	/**
	 * @var ITempManager
	 */
	private $tempManager;

	/**
	 * @param string $path
	 * @return string correctly encoded path
	 */
	private function normalizePath($path): string {
		$path = \trim($path, '/');

		if (!$path) {
			$path = '.';
		}

		return $path;
	}

	/**
	 * when running the tests wait to let the buckets catch up
	 */
	private function testTimeout(): void {
		if ($this->test) {
			\sleep($this->timeout);
		}
	}

	private function isRoot($path): bool {
		return $path === '.';
	}

	private function cleanKey($path): string {
		if ($this->isRoot($path)) {
			return '/';
		}
		return $path;
	}

	/**
	 * AmazonS3 constructor.
	 *
	 * @param array $params
	 *
	 * @throws \Exception
	 */
	public function __construct($params) {
		if (empty($params['key']) || empty($params['secret']) || empty($params['bucket'])) {
			throw new \Exception('Access Key, Secret and Bucket have to be configured.');
		}

		$this->id = 'amazon::' . $params['bucket'];

		$this->bucket = $params['bucket'];
		$this->test = isset($params['test']);
		$this->timeout = $params['timeout'] ?? 15;
		$this->rescanDelay = $params['rescanDelay'] ?? 10;
		$params['region'] = empty($params['region']) ? 'eu-west-1' : $params['region'];
		$params['hostname'] = empty($params['hostname']) ? 's3.amazonaws.com' : $params['hostname'];
		if (!isset($params['port']) || $params['port'] === '') {
			$params['port'] = ($params['use_ssl'] === false) ? 80 : 443;
		}
		$this->params = $params;
		$this->logger = \OC::$server->getLogger();
		$this->tempManager = \OC::$server->getTempManager();
	}

	/**
	 * Remove a file or folder
	 *
	 * @param string $path
	 * @return bool
	 * @throws StorageNotAvailableException
	 */
	protected function remove($path) {
		// remember fileType to reduce http calls
		$fileType = $this->filetype($path);
		if ($fileType === 'dir') {
			return $this->rmdir($path);
		}

		if ($fileType === 'file') {
			return $this->unlink($path);
		}

		return false;
	}

	public function mkdir($path) {
		$path = $this->normalizePath($path);

		if ($this->is_dir($path)) {
			return false;
		}

		try {
			$this->getConnection()->putObject([
				'Bucket' => $this->bucket,
				'Key' => $path . '/',
				'Body' => '',
				'ContentType' => 'httpd/unix-directory'
			]);
			$this->testTimeout();
		} catch (S3Exception $e) {
			$this->logger->logException($e, ['app'=>'files_external']);
			return false;
		}

		return true;
	}

	public function file_exists($path) {
		return $this->filetype($path) !== false;
	}

	public function rmdir($path) {
		$path = $this->normalizePath($path);

		if ($this->isRoot($path)) {
			return $this->clearBucket();
		}

		if (!$this->file_exists($path)) {
			return false;
		}

		return $this->batchDelete($path);
	}

	protected function clearBucket() {
		try {
			$this->getConnection()->deleteMatchingObjects($this->bucket);
			return true;
			// clearBucket() is not working with Ceph, so if it fails we try the slower approach
		} catch (\Exception $e) {
			return $this->batchDelete();
		}
	}

	/**
	 * @throws \Exception
	 */
	private function batchDelete($path = null) {
		$params = [
			'Bucket' => $this->bucket
		];
		if ($path !== null) {
			$params['Prefix'] = $path . '/';
		}
		try {
			// Since there are no real directories on S3, we need
			// to delete all objects prefixed with the path.
			do {
				// instead of the iterator, manually loop over the list ...
				$objects = $this->getConnection()->listObjects($params);
				/** @phan-suppress-next-line PhanDeprecatedFunction */
				$keys = $objects->getPath('Contents/*/Key');
				if (empty($keys)) {
					continue;
				}
				$keys = \array_map(static function ($key) {
					return ['Key' => $key];
				}, $keys);
				// ... so we can delete the files in batches
				$this->getConnection()->deleteObjects([
					'Bucket' => $this->bucket,
					'Delete' => [
						'Objects' => $keys
					]
				]);
				$this->testTimeout();
				// we reached the end when the list is no longer truncated
			} while ($objects['IsTruncated']);
		} catch (S3Exception $e) {
			$this->logger->logException($e, ['app'=>'files_external']);
			return false;
		}
		return true;
	}

	public function opendir($path) {
		$path = $this->normalizePath($path);

		if ($this->isRoot($path)) {
			$path = '';
		} else {
			$path .= '/';
		}

		try {
			$files = [];
			$result = $this->getConnection()->getPaginator('ListObjects', [
				'Bucket' => $this->bucket,
				'Delimiter' => '/',
				'Prefix' => $path
			]);

			foreach ($result->search('[CommonPrefixes[].Prefix, Contents[].Key][]') as $object) {
				if (\is_string($object)) {
					if ($object === $path) {
						// it's the directory itself, skip
						continue;
					}
					$file = \basename($object);
					$files[] = $file;
				}
			}

			return IteratorDirectory::wrap($files);
		} catch (S3Exception $e) {
			$this->logger->logException($e, ['app'=>'files_external']);
			return false;
		}
	}

	public function stat($path) {
		$path = $this->normalizePath($path);

		try {
			$stat = [];
			if ($this->is_dir($path)) {
				//folders don't really exist
				$stat['size'] = -1; //unknown
				$stat['mtime'] = \time() - $this->rescanDelay * 1000;
			} else {
				$result = $this->getConnection()->headObject([
					'Bucket' => $this->bucket,
					'Key' => $path
				]);

				$stat['size'] = $result['ContentLength'] ?: 0;
				if (isset($result['Metadata']['lastmodified'])) {
					// The underlying type doc for headObject says that it is an \Aws\Result
					// And phan understands that Result should have only 1 level
					// of array keys that are strings. But actually it can have 2 levels.
					/** @phan-suppress-next-line PhanTypeMismatchDimFetch */
					$stat['mtime'] = \strtotime($result['Metadata']['lastmodified']);
				} else {
					$stat['mtime'] = \strtotime($result['LastModified']);
				}
			}
			$stat['atime'] = \time();

			return $stat;
		} catch (S3Exception $e) {
			$this->logger->logException($e, ['app'=>'files_external']);
			return false;
		}
	}

	public function filetype($path) {
		$path = $this->normalizePath($path);

		if ($this->isRoot($path)) {
			return 'dir';
		}

		try {
			if ($this->getConnection()->doesObjectExist($this->bucket, $path)) {
				return 'file';
			}
			if ($this->getConnection()->doesObjectExist($this->bucket, $path.'/')) {
				return 'dir';
			}
		} catch (S3Exception $e) {
			$this->logger->logException($e, ['app'=>'files_external']);
			return false;
		}

		return false;
	}

	/**
	 * @throws StorageNotAvailableException
	 */
	public function unlink($path) {
		$path = $this->normalizePath($path);

		if ($this->is_dir($path)) {
			return $this->rmdir($path);
		}

		try {
			$this->getConnection()->deleteObject([
				'Bucket' => $this->bucket,
				'Key' => $path
			]);
			$this->testTimeout();
		} catch (S3Exception $e) {
			$this->logger->logException($e, ['app'=>'files_external']);
			return false;
		}

		return true;
	}

	public function fopen($path, $mode) {
		$path = $this->normalizePath($path);

		switch ($mode) {
			case 'r':
			case 'rb':
				$tmpFile = $this->tempManager->getTemporaryFile();
				self::$tmpFiles[$tmpFile] = $path;

				try {
					$this->getConnection()->getObject([
						'Bucket' => $this->bucket,
						'Key' => $path,
						'SaveAs' => $tmpFile
					]);
				} catch (S3Exception $e) {
					$this->logger->logException($e, ['app'=>'files_external']);
					return false;
				}

				return \fopen($tmpFile, 'rb');
			case 'w':
			case 'wb':
			case 'a':
			case 'ab':
			case 'r+':
			case 'w+':
			case 'wb+':
			case 'a+':
			case 'x':
			case 'x+':
			case 'c':
			case 'c+':
				if (\strrpos($path, '.') !== false) {
					$ext = \substr($path, \strrpos($path, '.'));
				} else {
					$ext = '';
				}
				$tmpFile = $this->tempManager->getTemporaryFile($ext);
				\OC\Files\Stream\Close::registerCallback($tmpFile, [$this, 'writeBack']);
				if ($this->file_exists($path)) {
					$source = $this->fopen($path, 'r');
					\file_put_contents($tmpFile, $source);
				}
				self::$tmpFiles[$tmpFile] = $path;

				return \fopen('close://' . $tmpFile, $mode);
		}
		return false;
	}

	public function touch($path, $mtime = null) {
		$path = $this->normalizePath($path);

		if ($mtime === null) {
			$mtime = \time();
		}
		$metadata = [
			'lastmodified' => \gmdate(DateTimeInterface::RFC1123, $mtime)
		];

		$fileType = $this->filetype($path);
		try {
			if ($fileType !== false) {
				if ($fileType === 'dir' && ! $this->isRoot($path)) {
					$path .= '/';
				}
				$this->getConnection()->copyObject([
					'Bucket' => $this->bucket,
					'Key' => $this->cleanKey($path),
					'Metadata' => $metadata,
					'CopySource' => $this->bucket . '/' . $path,
					'MetadataDirective' => 'REPLACE',
				]);
				$this->testTimeout();
			} else {
				$mimeType = \OC::$server->getMimeTypeDetector()->detectPath($path);
				$this->getConnection()->putObject([
					'Bucket' => $this->bucket,
					'Key' => $this->cleanKey($path),
					'Metadata' => $metadata,
					'Body' => '',
					'ContentType' => $mimeType,
					'MetadataDirective' => 'REPLACE',
				]);
				$this->testTimeout();
			}
		} catch (S3Exception $e) {
			$this->logger->logException($e, ['app'=>'files_external']);
			return false;
		}

		return true;
	}

	public function copy($path1, $path2) {
		$path1 = $this->normalizePath($path1);
		$path2 = $this->normalizePath($path2);

		if ($this->is_file($path1)) {
			try {
				$this->getConnection()->copyObject([
					'Bucket' => $this->bucket,
					'Key' => $this->cleanKey($path2),
					'CopySource' => S3Client::encodeKey($this->bucket . '/' . $path1)
				]);
				$this->testTimeout();
			} catch (S3Exception $e) {
				$this->logger->logException($e, ['app'=>'files_external']);
				return false;
			}
		} else {
			$this->remove($path2);

			try {
				$this->getConnection()->copyObject([
					'Bucket' => $this->bucket,
					'Key' => $path2 . '/',
					'CopySource' => S3Client::encodeKey($this->bucket . '/' . $path1 . '/')
				]);
				$this->testTimeout();
			} catch (S3Exception $e) {
				$this->logger->logException($e, ['app'=>'files_external']);
				return false;
			}

			$dh = $this->opendir($path1);
			if (\is_resource($dh)) {
				while (($file = \readdir($dh)) !== false) {
					if (\OC\Files\Filesystem::isIgnoredDir($file)) {
						continue;
					}

					$source = $path1 . '/' . $file;
					$target = $path2 . '/' . $file;
					$this->copy($source, $target);
				}
			}
		}

		return true;
	}

	public function rename($path1, $path2) {
		$path1 = $this->normalizePath($path1);
		$path2 = $this->normalizePath($path2);

		if ($this->is_file($path1)) {
			if ($this->copy($path1, $path2) === false) {
				return false;
			}

			if ($this->unlink($path1) === false) {
				$this->unlink($path2);
				return false;
			}
		} else {
			if ($this->copy($path1, $path2) === false) {
				return false;
			}

			if ($this->rmdir($path1) === false) {
				$this->rmdir($path2);
				return false;
			}
		}

		return true;
	}

	/**
	 * @throws \Exception
	 */
	public function test() {
		if ($this->getConnection()->getApi()->hasOperation('GetBucketAcl')) {
			$test = $this->getConnection()->getBucketAcl([
				'Bucket' => $this->bucket,
			]);
			/** @phan-suppress-next-line PhanDeprecatedFunction */
			if ($test !== null && $test->getPath('Owner/ID') !== null) {
				return true;
			}
			return false;
		}

		$buckets = $this->getConnection()->listBuckets();
		/** @phan-suppress-next-line PhanDeprecatedFunction */
		if ($buckets->getPath('Owner/ID') === null) {
			return false;
		}
		/** @phan-suppress-next-line PhanDeprecatedFunction */
		$bucketExists = !empty(\array_filter($buckets->getPath('Buckets'), function ($k) {
			return $k['Name'] === $this->bucket;
		}));
		return $bucketExists;
	}

	public function getId() {
		return $this->id;
	}

	/**
	 * Returns the connection
	 *
	 * @return S3Client connected client
	 * @throws \Exception if connection could not be made
	 */
	public function getConnection(): S3Client {
		if ($this->connection !== null) {
			return $this->connection;
		}

		$scheme = ($this->params['use_ssl'] === false) ? 'http' : 'https';
		$base_url = $scheme . '://' . $this->params['hostname'] . ':' . $this->params['port'] . '/';

		$config = [
			'version' => '2006-03-01',
			'region' => $this->params['region'],
//			'signature_version' => 'v4',
			'credentials' => [
				'key' => $this->params['key'],
				'secret' => $this->params['secret'],
			],
			'endpoint' => $base_url,
			'use_path_style_endpoint' => $this->params['use_path_style'],
		];
		// Create a handler stack that has all the default middlewares attached
		$handler = \GuzzleHttp\HandlerStack::create(new CurlMultiHandler());
		// Push the handler onto the handler stack
		$handler->push(Middleware::mapRequest(static function (RequestInterface $request) {
			if ($request->getMethod() !== 'PUT') {
				return $request;
			}
			$body = $request->getBody();
			if ($body !== null && $body->getSize() !== 0) {
				return $request;
			}
			if ($request->hasHeader('Content-Length')) {
				return $request;
			}
			// force content length header on empty body
			return $request->withHeader('Content-Length', '0');
		}));
		// Inject the handler into the client
		$client = new \GuzzleHttp\Client(['handler' => $handler]);
		$h = new GuzzleHandler($client);
		$config['http_handler'] = $h;

		$this->connection = new S3Client($config);

		if (!$this->connection->doesBucketExist($this->bucket)) {
			try {
				$this->connection->createBucket([
					'Bucket' => $this->bucket
				]);
				/* @phan-suppress-next-line PhanUndeclaredFunctionInCallable */
				$this->connection->waitUntil('BucketExists', [
					'Bucket' => $this->bucket,
					'waiter.interval' => 1,
					'waiter.max_attempts' => 15
				]);
				$this->testTimeout();
			} catch (S3Exception $e) {
				$this->logger->logException($e, ['app'=>'files_external']);
				throw new \Exception('Creation of bucket failed. '.$e->getMessage());
			}
		}

		return $this->connection;
	}

	/**
	 * @throws \Exception
	 */
	public function writeBack($tmpFile) {
		if (!isset(self::$tmpFiles[$tmpFile])) {
			return false;
		}

		try {
			$this->getConnection()->putObject([
				'Bucket' => $this->bucket,
				'Key' => $this->cleanKey(self::$tmpFiles[$tmpFile]),
				'SourceFile' => $tmpFile,
				'ContentType' => \OC::$server->getMimeTypeDetector()->detect($tmpFile),
				'ContentLength' => \filesize($tmpFile)
			]);
			$this->testTimeout();

			\unlink($tmpFile);
		} catch (S3Exception $e) {
			$this->logger->logException($e, ['app'=>'files_external']);
			return false;
		}
	}

	/**
	 * check if curl is installed
	 */
	public static function checkDependencies() {
		return true;
	}

	public function usePartFile() {
		return false;
	}
}
